packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js

```diff
@@ -0,0 +1,240 @@
+/**
+ * External dependencies
+ */
+import { invert, omit, keyBy, isEqual } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { serialize, parse, createBlock } from '@wordpress/blocks';
+import { useState, useEffect, useCallback, useRef } from '@wordpress/element';
+
+function blockToWidget( block, existingWidget = null ) {
+	let widget;
+
+	if ( block.name === 'core/legacy-widget' ) {
+		const isReferenceWidget = !! block.attributes.referenceWidgetName;
+		if ( isReferenceWidget ) {
+			widget = {
+				id: block.attributes.referenceWidgetName,
+				instance: block.attributes.instance,
+			};
+		} else {
+			widget = {
+				widgetClass: block.attributes.widgetClass,
+				idBase: block.attributes.idBase,
+				instance: block.attributes.instance,
+			};
+		}
+	} else {
+		widget = {
+			idBase: 'block',
+			widgetClass: 'WP_Widget_Block',
+			instance: { content: serialize( block ) },
+		};
+	}
+
+	return {
+		...omit( existingWidget, [ 'form', 'rendered' ] ),
+		...widget,
+	};
+}
+
+function widgetToBlock( widget, existingBlock = null ) {
+	let block;
+
+	// FIXME: We'll never get it here with blocks, we need to update this.
+	if ( widget.widgetClass === 'WP_Widget_Block' ) {
+		const parsedBlocks = parse( widget.instance.content );
+		block = parsedBlocks.length
+			? parsedBlocks[ 0 ]
+			: createBlock( 'core/paragraph', {} );
+	} else {
+		const attributes = {
+			name: widget.name,
+			form: widget.form,
+			instance: widget.instance,
+			idBase: widget.idBase,
+			number: widget.number,
+		};
+
+		const isReferenceWidget = ! widget.widgetClass;
+		if ( isReferenceWidget ) {
+			attributes.referenceWidgetName = widget.id;
+		} else {
+			attributes.widgetClass = widget.widgetClass;
+		}
+
+		block = createBlock( 'core/legacy-widget', attributes, [] );
+	}
+
+	return {
+		...block,
+		clientId: existingBlock?.clientId ?? block.clientId,
+	};
+}
+
+function initState( sidebar ) {
+	const state = { blocks: [], widgetIds: {} };
+
+	for ( const widgetId of sidebar.getWidgetIds() ) {
+		const widget = sidebar.getWidget( widgetId );
+		const block = widgetToBlock( widget );
+		state.blocks.push( block );
+		state.widgetIds[ block.clientId ] = widgetId;
```

Comment:
Can probably simplify how this hook works using `__internalWidgetId` similar to https://github.com/WordPress/gutenberg/pull/28379. Not right now though.